File: packages/python/elizaos/features/advanced_capabilities/actions/roles.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import UPDATE_ROLE_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_action_spec("UPDATE_ROLE")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    spec_examples = _spec.get("examples", [])
    if not isinstance(spec_examples, list):
        return []
    result: list[list[ActionExample]] = []
    for example in spec_examples:
        if not isinstance(example, list):
            continue
        row: list[ActionExample] = []
        for msg in example:
            if not isinstance(msg, dict):
                continue
            content = msg.get("content", {})
            text = ""
            actions: list[str] | None = None
            if isinstance(content, dict):
                text_val = content.get("text", "")
                text = str(text_val) if text_val else ""
                actions_val = content.get("actions")
                if isinstance(actions_val, list) and all(isinstance(a, str) for a in actions_val):
                    actions = list(actions_val)
            row.append(
                ActionExample(
                    name=str(msg.get("name", "")),
                    content=Content(text=text, actions=actions),
                )
            )
        if row:
            result.append(row)
    return result


class Role(StrEnum):
    OWNER = "OWNER"
    ADMIN = "ADMIN"
    MEMBER = "MEMBER"
    GUEST = "GUEST"
    NONE = "NONE"


@dataclass
class UpdateRoleAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None or room.world_id is None:
            return False

        world = await runtime.get_world(room.world_id)
        if world is None or world.metadata is None:
            return False

        roles = world.metadata.get("roles", {})
        agent_role = roles.get(str(runtime.agent_id), Role.NONE.value)

        return agent_role in (Role.OWNER.value, Role.ADMIN.value)

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for UPDATE_ROLE action")

        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room context for role update",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None or room.world_id is None:
            return ActionResult(
                text="Room or world not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        world = await runtime.get_world(room.world_id)
        if world is None or world.metadata is None:
            return ActionResult(
                text="World not found",
                values={"success": False, "error": "world_not_found"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        state = await runtime.compose_state(
            message, ["RECENT_MESSAGES", "ACTION_STATE", "WORLD_INFO"]
        )

        current_roles = world.metadata.get("roles", {})
        roles_context = "\n".join(
            f"- {entity_id}: {role}" for entity_id, role in current_roles.items()
        )

        template = (
            runtime.character.templates.get("updateRoleTemplate")
            if runtime.character.templates and "updateRoleTemplate" in runtime.character.templates
            else UPDATE_ROLE_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)
        prompt = prompt.replace("{{roles}}", roles_context)

        response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response_text)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response")

        thought = str(parsed_xml.get("thought", ""))
        entity_id_str = str(parsed_xml.get("entity_id", ""))
        new_role_str = str(parsed_xml.get("new_role", "")).upper()

        if not entity_id_str:
            raise ValueError("No entity ID provided")

        if new_role_str not in [r.value for r in Role]:
            raise ValueError(f"Invalid role: {new_role_str}")

        entity_id = UUID(entity_id_str)

        roles = dict(world.metadata.get("roles", {}))
        old_role = roles.get(str(entity_id), Role.NONE.value)
        roles[str(entity_id)] = new_role_str
        world.metadata["roles"] = roles

        await runtime.update_world(world)

        response_content = Content(
            text=f"Updated role for {entity_id_str}: {old_role} -> {new_role_str}",
            actions=["UPDATE_ROLE"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Role updated: {entity_id_str} is now {new_role_str}",
            values={
                "success": True,
                "roleUpdated": True,
                "entityId": str(entity_id),
                "oldRole": old_role,
                "newRole": new_role_str,
            },
            data={
                "actionName": "UPDATE_ROLE",
                "entityId": str(entity_id),
                "oldRole": old_role,
                "newRole": new_role_str,
                "thought": thought,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


update_role_action = Action(
    name=UpdateRoleAction.name,
    similes=UpdateRoleAction().similes,
    description=UpdateRoleAction.description,
    validate=UpdateRoleAction().validate,
    handler=UpdateRoleAction().handler,
    examples=UpdateRoleAction().examples,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/actions/roles.rs
//! UPDATE_ROLE action implementation.

use async_trait::async_trait;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::{PluginError, PluginResult};
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{ActionResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Action;
use crate::generated::spec_helpers::require_action_spec;
use crate::prompts::UPDATE_ROLE_TEMPLATE;
use once_cell::sync::Lazy;

/// Action for updating entity roles.
pub struct UpdateRoleAction;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ActionDoc> =
    Lazy::new(|| require_action_spec("UPDATE_ROLE"));

#[async_trait]
impl Action for UpdateRoleAction {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn similes(&self) -> &[&'static str] {
        static SIMILES: Lazy<Box<[&'static str]>> = Lazy::new(|| {
            SPEC.similes
                .iter()
                .map(|s| s.as_str())
                .collect::<Vec<_>>()
                .into_boxed_slice()
        });
        &SIMILES
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        let Some(room_id) = message.room_id else {
            return false;
        };

        // Check if room exists and has a world
        let Ok(Some(room)) = runtime.get_room(room_id).await else {
            return false;
        };

        let Some(world_id) = room.world_id else {
            return false;
        };

        // Check if agent has permission to update roles
        let Ok(Some(world)) = runtime.get_world(world_id).await else {
            return false;
        };

        if let Some(metadata) = world.metadata.as_ref() {
            let agent_id = runtime.agent_id().to_string();
            if let Some(role) = metadata.roles.get(&agent_id) {
                return role == "OWNER" || role == "ADMIN";
            }
        }

        false
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let _state = state.ok_or_else(|| {
            PluginError::StateRequired("State is required for UPDATE_ROLE action".to_string())
        })?;

        let room_id = message.room_id.ok_or_else(|| {
            PluginError::InvalidInput("No room context for role update".to_string())
        })?;

        let room = runtime
            .get_room(room_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Room not found".to_string()))?;

        let world_id = room
            .world_id
            .ok_or_else(|| PluginError::InvalidInput("Room has no world".to_string()))?;

        let world = runtime
            .get_world(world_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("World not found".to_string()))?;

        // Compose state
        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES", "ACTION_STATE", "WORLD_INFO"])
            .await?;

        let roles_context = world
            .metadata
            .get("roles")
            .and_then(|roles| roles.as_object())
            .map(|roles| {
                roles
                    .iter()
                    .map(|(entity_id, role)| {
                        format!("- {}: {}", entity_id, role.as_str().unwrap_or("NONE"))
                    })
                    .collect::<Vec<_>>()
                    .join("\n")
            })
            .unwrap_or_default();

        // Get template
        let template = runtime
            .character()
            .templates
            .get("updateRoleTemplate")
            .map(|s| s.as_str())
            .unwrap_or(UPDATE_ROLE_TEMPLATE);

        let prompt = runtime
            .compose_prompt(&composed_state, template)
            .replace("{{roles}}", &roles_context);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse response XML".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let entity_id_str = parsed
            .get("entity_id")
            .cloned()
            .ok_or_else(|| PluginError::InvalidInput("No entity ID provided".to_string()))?;
        let new_role_str = parsed
            .get("new_role")
            .cloned()
            .ok_or_else(|| PluginError::InvalidInput("No role provided".to_string()))?
            .to_uppercase();

        // Validate entity ID
        let entity_id = Uuid::parse_str(&entity_id_str).map_err(|_| {
            PluginError::InvalidInput(format!("Invalid entity ID: {}", entity_id_str))
        })?;

        // Validate role
        match new_role_str.as_str() {
            "OWNER" | "ADMIN" | "MEMBER" | "GUEST" | "NONE" => {}
            _ => {
                return Err(PluginError::InvalidInput(format!(
                    "Invalid role: {}",
                    new_role_str
                )))
            }
        }

        // Get old role
        let old_role = world
            .metadata
            .as_ref()
            .and_then(|m| m.roles.get(&entity_id_str).cloned())
            .unwrap_or_else(|| "NONE".to_string());

        // Update role in world
        let mut updated_world = world.clone();
        let metadata = updated_world.metadata.get_or_insert_with(Default::default);
        metadata
            .roles
            .insert(entity_id_str.clone(), new_role_str.clone());
        runtime.update_world(&updated_world).await?;

        Ok(ActionResult::success(format!(
            "Role updated: {} is now {}",
            entity_id_str, new_role_str
        ))
        .with_value("success", true)
        .with_value("roleUpdated", true)
        .with_value("entityId", entity_id_str.clone())
        .with_value("oldRole", old_role.clone())
        .with_value("newRole", new_role_str.clone())
        .with_data("actionName", "UPDATE_ROLE")
        .with_data("entityId", entity_id_str)
        .with_data("oldRole", old_role)
        .with_data("newRole", new_role_str)
        .with_data("thought", thought))
    }
}
